connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java

```diff
@@ -58,41 +61,40 @@
 
     private Map<String, String> config;
     private State state;
-
-    public WorkerConnector(String connName,
-                           Connector connector,
-                           ConnectorContext ctx,
-                           ConnectMetrics metrics,
-                           ConnectorStatus.Listener statusListener) {
+    private final OffsetStorageReader offsetStorageReader;
+
+    public WorkerConnector(final String connName,
+                           final Connector connector,
+                           final ConnectorContext ctx,
+                           final ConnectMetrics metrics,
+                           final ConnectorStatus.Listener statusListener,
+                           final OffsetStorageReader offsetStorageReader) {
         this.connName = connName;
         this.ctx = ctx;
         this.connector = connector;
         this.state = State.INIT;
         this.metrics = new ConnectorMetricsGroup(metrics, AbstractStatus.State.UNASSIGNED, statusListener);
         this.statusListener = this.metrics;
+        this.offsetStorageReader = offsetStorageReader;
     }
 
     public void initialize(ConnectorConfig connectorConfig) {
         try {
             this.config = connectorConfig.originalsStrings();
+
             log.debug("{} Initializing connector {}", this, connName);
             if (isSinkConnector()) {
                 SinkConnectorConfig.validate(config);
             }
 
-            connector.initialize(new ConnectorContext() {
-                @Override
-                public void requestTaskReconfiguration() {
-                    ctx.requestTaskReconfiguration();
-                }
-
-                @Override
-                public void raiseError(Exception e) {
-                    log.error("{} Connector raised an error", WorkerConnector.this, e);
-                    onFailure(e);
-                    ctx.raiseError(e);
-                }
-            });
+            final ConnectorContext delegateCtx = new DelegateToWorkerConnectorContext();
+
+            if (isSinkConnector()) {
+                SinkConnectorConfig.validate(config);
+                connector.initialize(new DelegateSinkConnectorContext(delegateCtx));
+            } else {
+                connector.initialize(new DelegateSourceConnectorContext(delegateCtx, offsetStorageReader));
+            }
```

Comment:
At one point these were anonymous classes, and the delegation perhaps made a bit more sense. Now that they are inner classes, it seems like it would be a lot less confusing and simpler if the `DelegateToWorkerConnectorContext` class were extended by the `DelegateSinkConnectorContext` and `DelegateSourceConnectorContext` classes. Doing this has several advantages:
1. removes nesting/delegation and eliminates the chance of getting into an infinite loop
2. eliminates duplicate code in these classes
3. simplifies the context classes quite a bit
4. makes it more obvious that the sink connector context has nothing extra over the base
5. makes it more obvious that the source connector context only adds the offset storage reader
6. simplifies this code a bit (see below)

By keeping `DelegateToWorkerConnectorContext` class non-abstract, we're actually able to maintain backward compatibility by calling initialize when the connector class is neither a source or sink connector:

This code becomes simpler, too:
```suggestion
            if (isSinkConnector()) {
                SinkConnectorConfig.validate(config);
                connector.initialize(new DelegateSinkConnectorContext());
            } else if (isSourceConnector()) {
                connector.initialize(new DelegateSourceConnectorContext(offsetStorageReader));
            } else {
                connector.initialize(new DelegateToWorkerConnectorContext());
            }
```
This seems a little strange, but we're actually not checking whether the `Connector` instance passed into this `WorkerConnector` is only an instance of `SourceConnector` or `SinkConnector`. If it is neither, prior to this change the framework would still initialize it, and I think we should maintain that arguably strange behavior just to maintain backward compatibility.